### PR TITLE
ZIP

### LINKED ISSUE #101: Rewrite unzipping logic with libraries instead of 7z
# Description

Currently the core downloads and invoke 7z to handle zipping and unzipping. This is not ideal for multiple reasons.

Rewrite the unzipping logic to use libraries.

https://rust-lang-nursery.github.io/rust-cookbook/compression/tar.html

https://github.com/google/ripunzip

https://github.com/muja/unrar.rs

### PR DESCRIPTION
**Pull request by jasonw930**
# Description

ZIP

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

*Note: make sure your files are formatted with rust-analyzer*

### CODE REVIEW

**Review by CheatCod — CHANGES_REQUESTED** [review]
Some test cases would be nice :)

**Comment by CheatCod** [comment]
There is a bug involved in the old unzip logic, where the function does a diff of the destination dir before and after the unzip. So if the unzip overwrote some files, the file paths are not returned.

The desired behavior will be to get the content of the archive (listing the files in the temp dir)

File: src/util.rs [hunk 1]
```diff
@@ -158,81 +161,79 @@ pub async fn unzip_file(
...
+        archive.unpack(temp_dest).context(format!("Failed to decompress file {}", file.display()))?;
+    } else if file_extension == "zip" {
+        let zip = std::fs::File::open(file).context(format!("Failed to open file {}", file.display()))?;
+        let mut archive = zip::ZipArchive::new(zip).context(format!("Failed to decompress file {}", file.display()))?;
+        archive.extract(temp_dest).context(format!("Failed to decompress file {}", file.display()))?;
+    } else if file_extension == "rar" {
+        let archive = unrar::Archive::new(file.to_str().ok_or_else(|| eyre!("Non-unicode character in file name {}", file.display()))?.to_string());
+        archive
+            .extract_to(temp_dest.to_str().ok_or_else(|| eyre!("Non-unicode character in file name {}", temp_dest.display()))?.to_string())
+            .map_err(|_| eyre!("Failed to decompress file {}", file.display()))?
```
Comment (CheatCod) [inline src/util.rs:219]: use .context

File: src/util.rs [hunk 2]
```diff
@@ -158,81 +161,79 @@ pub async fn unzip_file(
...
+            .process()
+            .map_err(|_| eyre!("Failed to decompress file {}", file.display()))?;
```
Comment (CheatCod) [inline src/util.rs:221]: .context

**Review by CheatCod — CHANGES_REQUESTED** [review]

**Review by CheatCod — CHANGES_REQUESTED** [review]
Add test case for .tar.gz

File: src/util.rs (hunk 2 above)
Comment (jasonw930) [reply src/util.rs:221]: note: the following trait bounds were not satisfied:
            `UnrarError<Vec<unrar::archive::Entry>>: eyre::context::ext::StdError`

They didn't make their error type good I guess

**Review by CheatCod — APPROVED** [review]
Please install clippy and watch out for minor details such as unused deps

**Pull request merged**